include/fiction/networks/virtual_pi_network.hpp

```diff
@@ -0,0 +1,693 @@
+     *
+     * @return The number of virtual CIs as a uint32_t.
+     */
+    [[nodiscard]] auto num_cis_virtual() const
+    {
+        return static_cast<uint32_t>(virtual_inputs->size());
+    }
+
+    /**
+     * @brief Get the number of real constant inputs (CIs) in the virtual_pi_network.
+     *
+     * The num_cis_real() function returns the total number of real CIs in the virtual_pi_network.
+     * Real constant inputs are those which exist in the inputs vector of the `_storage` object.
+     *
+     * @return The number of real CIs as a uint32_t.
+     */
+    [[nodiscard]] auto num_cis_real() const
+    {
+        return static_cast<uint32_t>(_storage->inputs.size());
+    }
+
+    /**
+     * @brief Reimplementation of num_pis(). Get the number of primary inputs (PIs) in the virtual_pi_network.
+     *
+     * The num_pis() function returns the total number of PIs in the virtual_pi_network.
+     * It includes both virtual PIs and real PIs.
+     *
+     * @return The number of PIs as a uint32_t.
+     */
+    [[nodiscard]] auto num_pis() const
+    {
+        return static_cast<uint32_t>(_storage->inputs.size() + virtual_inputs->size());
+    }
+
+    /**
+     * @brief Get the number of virtual primary inputs (PIs) in the virtual_pi_network.
+     *
+     * The num_pis_virtual() function returns the total number of virtual PIs in the virtual_pi_network.
+     * Virtual PIs are those which exist in the virtual_inputs vector.
+     *
+     * @return The number of virtual PIs as a uint32_t.
+     */
+    [[nodiscard]] auto num_pis_virtual() const
+    {
+        return static_cast<uint32_t>(virtual_inputs->size());
+    }
+
+    /**
+     * @brief Retrieve the number of real primary inputs (PIs) in the virtual_pi_network.
+     *
+     * The num_pis_real() function returns the total number of real PIs in the virtual_pi_network.
+     * Real PIs are those which exist in the inputs vector of the `_storage` object.
+     *
+     * @return The number of real PIs as a uint32_t.
+     */
+    [[nodiscard]] auto num_pis_real() const
+    {
+        return static_cast<uint32_t>(_storage->inputs.size());
+    }
+
+    /**
+     * @brief Retrieve the real PI mapped to the given virtual PI.
+     *
+     * This function searches for the given virtual PI in the mapping table and returns the corresponding real PI.
+     * It throws a std::runtime_error if the virtual PI is not found in the mapping table.
+     *
+     * @param v_pi The virtual PI to find the corresponding real PI for.
+     * @return The real PI mapped to the virtual PI.
+     * @throws std::runtime_error If the virtual PI is not found in the mapping table.
+     */
+    [[nodiscard]] auto get_real_pi(const node& v_pi) const
+    {
+        auto it = std::find_if(map.begin(), map.end(), [v_pi](const auto& pair) { return pair.first == v_pi; });
+
+        if (it != map.end())
+            return it->second;
```

Comment:
```suggestion
        if (it != map.end())
        {
            return it->second;
        }
```
Always wrap your statements. Not doing it, is bug-prone and has led to critical bugs in the real world.